### PR TITLE
Add group_points_by_bounding_boxes data augmentation.

### PR DESCRIPTION
**Pull request by lengzq**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [x] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by lengzq** [comment]
@tanzhenyu

File: keras_cv/layers/preprocessing3d/crop_objects.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        The first 5 features are [x, y, z, class, range].
+      bounding_boxes: 3D (multi frames) float32 Tensor with shape
+        [num of frames, num of boxes, num of box features].
+        The first 8 features are [x, y, z, dx, dy, dz, phi, box class].
+
+    Output shape:
+      A dictionary of Tensors with the same shape as input Tensors and two additional items for object_point_clouds and OBJECT_BOUNDING_BOXES.
+
+    Arguments:
+      label_index: A int scalar sets the target object index.
```
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing3d/crop_objects.py:45]: this still seems to generic, can you elaborate on what "label_index" mean? for example, crop all objects that does not equal to `label_index`

File: keras_cv/layers/preprocessing3d/crop_objects.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        points_in_bounding_boxes = is_within_box3d(
+            point_clouds, object_bounding_boxes[:, :, :7]
+        )
+        # Filter bounding boxes using the current frame.
+        min_points_filter = (
+            tf.reduce_sum(tf.cast(points_in_bounding_boxes[0], dtype=tf.int32), axis=0)
+            >= self._min_points_per_bounding_boxes
+        )
+
+        object_bounding_boxes = tf.boolean_mask(
```
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing3d/crop_objects.py:99]: to me, this is still bounding_boxes, but it's a filtered version of bounding_boxes, no? (I was trying to understand why we have two different formats, one is BOUNDING_BOXES, the other one is OBJECT_BOUNDING_BOXES

File: keras_cv/layers/preprocessing3d/crop_objects.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+BOUNDING_BOXES = base_augmentation_layer_3d.BOUNDING_BOXES
+OBJECT_POINT_CLOUDS = base_augmentation_layer_3d.OBJECT_POINT_CLOUDS
+OBJECT_BOUNDING_BOXES = base_augmentation_layer_3d.OBJECT_BOUNDING_BOXES
+
+
+class CropObjects(base_augmentation_layer_3d.BaseAugmentationLayer3D):
+    """A preprocessing layer which crops objects from point clouds during training.
+
+    This layer will crop points based on object bounding boxes and generates OBJECT_POINT_CLOUDS and OBJECT_BOUNDING_BOXES tensors.
+    During inference time, the output will be identical to input. Call the layer with `training=True` to crop the input points.
```
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing3d/crop_objects.py:31]: given that during training time, this changes from point_cloud to object_point_cloud, I assume the layers that follow this layer will need to accept object_point_cloud. In that case, does returning point_cloud during inference time make sense?

**Review by tanzhenyu — COMMENTED** [review]
Thanks for the PR!

File: keras_cv/layers/preprocessing3d/crop_objects.py (hunk 2 above)
Comment (lengzq) [reply keras_cv/layers/preprocessing3d/crop_objects.py:99]: OBJECT_BOUNDING_BOXES is a subset of BOUNDING_BOXES

File: keras_cv/layers/preprocessing3d/crop_objects.py (hunk 3 above)
Comment (lengzq) [reply keras_cv/layers/preprocessing3d/crop_objects.py:31]: The following layers will accept both object_point_cloud and the original point_cloud. I think we need to return point_cloud as well

File: keras_cv/layers/preprocessing3d/crop_objects.py (hunk 2 above)
Comment (lengzq) [reply keras_cv/layers/preprocessing3d/crop_objects.py:99]: Yes, it is a filtered version of BOUNDING_BOXES

Comment (tanzhenyu) [reply keras_cv/layers/preprocessing3d/crop_objects.py]: so, can we re-use BOUNDING_BOXES?

File: keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    Input shape:
+      point_clouds: 3D (multi frames) float32 Tensor with shape
+        [num of frames, num of points, num of point features].
+        The first 5 features are [x, y, z, class, range].
+      bounding_boxes: 3D (multi frames) float32 Tensor with shape
+        [num of frames, num of boxes, num of box features].
+        The first 8 features are [x, y, z, dx, dy, dz, phi, box class].
+
+    Output shape:
+      A dictionary of Tensors with the same shape as input Tensors and two additional items for OBJECT_POINT_CLOUDS and OBJECT_BOUNDING_BOXES.
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py:42]: Maybe let's include the shape of OBJECT_POINT_CLOUDS and OBJECT_BOUNDING_BOXES here?

File: keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+    Arguments:
+      label_index: A int scalar sets the target object index. Bounding boxes and corresponding point clouds with box class == label_index will be saved as OBJECT_BOUNDING_BOXES and OBJECT_POINT_CLOUDS.
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py:45]: Based on the implementation, it seems like 0 is a special case for this parameter where we include all boxes with non-zero classes, right?

If so, let's document that here.

File: keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes_test.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+)
+
+POINT_CLOUDS = base_augmentation_layer_3d.POINT_CLOUDS
+BOUNDING_BOXES = base_augmentation_layer_3d.BOUNDING_BOXES
+OBJECT_POINT_CLOUDS = base_augmentation_layer_3d.OBJECT_POINT_CLOUDS
+OBJECT_BOUNDING_BOXES = base_augmentation_layer_3d.OBJECT_BOUNDING_BOXES
+
+
+class GroupPointsByBoundingBoxesTest(tf.test.TestCase):
+    def test_augment_point_clouds_and_bounding_boxes(self):
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes_test.py:29]: It's probably worth including a test case which demonstrates that:

- calling the layer with training=False is a noop
- the layer preserves the input point clouds and bounding boxes (and potentially any other keys thrown at it)

File: keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py (hunk 5 above)
Comment (ianstenbit) [reply keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py:45]: Also, if we're special casing this, I think it's probably more intuitive to a user to have the default be `None`, and have that reflect all classes (instead of 0, which seems a little bit like it's a real class ID and not a sentinel value)

**Review by ianstenbit — CHANGES_REQUESTED** [review]

**Comment by ianstenbit** [comment]
thanks for the PR!

File: keras_cv/layers/preprocessing3d/crop_objects.py (hunk 3 above)
Comment (lengzq) [reply keras_cv/layers/preprocessing3d/crop_objects.py]: No augmentation at all during inference.

File: keras_cv/layers/preprocessing3d/crop_objects.py (hunk 2 above)
Comment (lengzq) [reply keras_cv/layers/preprocessing3d/crop_objects.py]: I think it is better to use different names to distinguish OBJECT_BOUNDING_BOXES from BOUNDING_BOXES. Same for POINT_CLOUDS and OBJECT_POINT_CLOUDS.

File: keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py (hunk 4 above)
Comment (lengzq) [reply keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py:42]: Done.

File: keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py (hunk 5 above)
Comment (lengzq) [reply keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py:45]: Done. I set default label_index = None.

File: keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes_test.py (hunk 6 above)
Comment (lengzq) [reply keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes_test.py:29]: 1. Done.
2. Do you mean the following?
```
        self.assertAllClose(inputs[POINT_CLOUDS], outputs[POINT_CLOUDS])
        self.assertAllClose(inputs[BOUNDING_BOXES], outputs[BOUNDING_BOXES])
```

Comment (ianstenbit) [reply keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes_test.py:29]: Yes I think that would be reasonable (and maybe also passing some other random key as an input to the layer to make sure it gets preserved in the output)

File: keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+        )
+        object_point_clouds = tf.gather(
+            object_point_clouds, sort_valid_index, axis=2, batch_dims=2
+        )[:, :, : self._max_points_per_bounding_boxes, :]
+
+        object_point_clouds = tf.where(
+            sort_valid_mask[:, :, :, tf.newaxis] > 0, object_point_clouds, 0.0
+        )
+
+        return (
```
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py:129]: following up on the bb vs object_bb -- the output of this layer, or the input to all following layers, should it be a dict, and should the keys be "point_cloud" and "bounging_boxes"?

Comment (lengzq) [reply keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py:129]: Yes, I think the output of this layer should be a dictionary of {'point_clouds': tf.Tensor, 'bounding_boxes': tf.Tensor, 'object_point_clouds': tf.Tensor, 'object_bounding_boxes': tf.Tensor}

File: keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        points_in_bounding_boxes = tf.boolean_mask(
+            points_in_bounding_boxes, min_points_filter, axis=2
+        )
+        # [num of frames, num of boxes, num of points].
+        points_in_bounding_boxes = tf.transpose(points_in_bounding_boxes, [0, 2, 1])
+        points_in_bounding_boxes = tf.cast(points_in_bounding_boxes, tf.int32)
+        sort_valid_index = tf.argsort(
+            points_in_bounding_boxes, axis=-1, direction="DESCENDING"
+        )
+        sort_valid_mask = tf.gather(
```
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py:113]: for padding -- shouldn't it pad with all zeros, instead of padding with points that are not within the box?

File: keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+        result = {
+            OBJECT_POINT_CLOUDS: object_point_clouds,
+            OBJECT_BOUNDING_BOXES: object_bounding_boxes,
+        }
+
+        # preserve any additional inputs unmodified by this layer.
+        for key in inputs.keys() - result.keys():
+            result[key] = inputs[key]
+        return result
+
```
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py:161]: nit: maybe you can start with results = inputs, and then results.update({OBJECT_POINT_CLOUD, OBJECT_BOUNDING_BOXES})

File: keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+    #             )
+    #             return inputs
+    #         else:
+    #             raise ValueError(
+    #                 "Point clouds augmentation layers are expecting inputs point clouds and bounding boxes to "
+    #                 "be rank 3D (Frame, Point, Feature) or 4D (Batch, Frame, Point, Feature) tensors. Got shape: {} and {}".format(
+    #                     point_clouds.shape, bounding_boxes.shape
+    #                 )
+    #             )
+    #     else:
```
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py:198]: does the commented out branch need to be updated for the new output format?

**Review by tanzhenyu — APPROVED** [review]

**Comment by tanzhenyu** [comment]
also the formatting test is failing

File: keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py (hunk 8 above)
Comment (lengzq) [reply keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py:113]: Yes, padded object_point_clouds is replaced with zero before return, i.e.,

```
        object_point_clouds = tf.where(
            sort_valid_mask[:, :, :, tf.newaxis] > 0, object_point_clouds, 0.0
        )
```

**Review by tanzhenyu — APPROVED** [review]

File: keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+        The first 5 features are [x, y, z, class, range].
+      bounding_boxes: 3D (multi frames) float32 Tensor with shape
+        [num of frames, num of boxes, num of box features].
+        The first 8 features are [x, y, z, dx, dy, dz, phi, box class].
+
+    Output shape:
+      A dictionary of Tensors with the same shape as input Tensors and two additional items for OBJECT_POINT_CLOUDS (shape [num of frames, num of valid boxes, max num of points, num of point features]) and OBJECT_BOUNDING_BOXES (shape [num of frames, num of valid boxes, num of box features]).
+
+    Arguments:
+      label_index: An optional int scalar sets the target object index. Bounding boxes and corresponding point clouds with box class == label_index will be saved as OBJECT_BOUNDING_BOXES and OBJECT_POINT_CLOUDS. If label index is None, all valid boundinhg boxes (box class !=0) are used.
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py:45]: nit: s/boundinhg/bounding

File: keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes_test.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+            [[[[0, 1, 2, 3, 4], [0, -1, 2, 3, 4]], [[10, 1, 2, 3, 4], [0, 0, 0, 0, 0]]]]
+            * 2
+        ).astype("float32")
+        object_bounding_boxes = np.array(
+            [[[0, 0, 0, 4, 4, 4, 0, 1], [10, 1, 2, 2, 2, 2, 0, 1]]] * 2
+        ).astype("float32")
+        self.assertAllClose(inputs[POINT_CLOUDS], outputs[POINT_CLOUDS])
+        self.assertAllClose(inputs[BOUNDING_BOXES], outputs[BOUNDING_BOXES])
+        self.assertAllClose(inputs["dummy_item"], outputs["dummy_item"])
+        # Sort the point clouds due to the orders of points are different when using Tensorflow and Metal+Tensorflow (MAC).
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes_test.py:65]: just curious -- is this a known behavior? Should this be an issue we open with TF?

**Review by ianstenbit — APPROVED** [review]
Great -- thank you!

File: keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes_test.py (hunk 12 above)
Comment (lengzq) [reply keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes_test.py:65]: Not to my knowledge.
Here is a simple example:
```
 tf.argsort([1, 1, 0], direction="DESCENDING")

TF on mac
 <tf.Tensor: shape=(3,), dtype=int32, numpy=array([1, 0, 2], dtype=int32)>
TF on linux 
<tf.Tensor: shape=(3,), dtype=int32, numpy=array([0, 1, 2], dtype=int32)>
```

File: keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py (hunk 11 above)
Comment (lengzq) [reply keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py]: Done.

File: keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py (hunk 10 above)
Comment (lengzq) [reply keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py]: Uncommented

File: keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py (hunk 9 above)
Comment (lengzq) [reply keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py]: Done.
